docs/lang/articles/misc/ndarray_android.md

```diff
@@ -0,0 +1,449 @@
+
+:::
+
+### Generate necessary compute shaders
+
+The following Python script defines a Taichi AOT module for generating and saving the necessary compute shaders (GLES shaders in this case) based on the chosen backend (OpenGL).
+
+
+
+```Python
+import taichi as ti
+
+ti.init(arch=ti.opengl, use_gles=True, allow_nv_shader_extension=False)
+
+# Define constants for computation
+G = 1
+PI = 3.141592653
+N = 1000
+m = 5
+galaxy_size = 0.4
+planet_radius = 1
+init_vel = 120
+h = 1e-5
+substepping = 10
+
+# Define Taichi kernels
+@ti.kernel
+def initialize(pos: ti.any_arr(element_dim=1), vel: ti.any_arr(element_dim=1)):
+    center=ti.Vector([0.5, 0.5])
+    for i in pos:
+        theta = ti.random() * 2 * PI
+        r = (ti.sqrt(ti.random()) * 0.7 + 0.3) * galaxy_size
+        offset = r * ti.Vector([ti.cos(theta), ti.sin(theta)])
+        pos[i] = center+offset
+        vel[i] = [-offset.y, offset.x]
+        vel[i] *= init_vel
+
+@ti.kernel
+def compute_force(pos: ti.any_arr(element_dim=1), vel: ti.any_arr(element_dim=1), force: ti.any_arr(element_dim=1)):
+    for i in pos:
+        force[i] = ti.Vector([0.0, 0.0])
+    for i in pos:
+        p = pos[i]
+        for j in pos:
+            if i != j:
+                diff = p-pos[j]
+                r = diff.norm(1e-5)
+                f = -G * m * m * (1.0/r)**3 * diff
+                force[i] += f
+    dt = h/substepping
+    for i in pos:
+        vel[i].atomic_add(dt*force[i]/m)
+        pos[i].atomic_add(dt*vel[i])
+
+# Define Ndarrays
+pos = ti.Vector.ndarray(2, ti.f32, N)
+vel = ti.Vector.ndarray(2, ti.f32, N)
+force = ti.Vector.ndarray(2, ti.f32, N)
+
+# Run the AOT module builder
+def aot():
+    m = ti.aot.Module(ti.opengl)
+    m.add_kernel(initialize, (pos, vel))
+    m.add_kernel(compute_force, (pos, vel, force))
+
+    dir_name = 'nbody_aot'
+    m.save(dir_name, '')
+aot()
+```
+
+**In line 3, you initialize Taichi:**
+
+1. Set `use_gles` to `True` to generate GLES compute shaders on Android.
+2. Set `allow_nv_shader_extension` to `False` for the GLES compute shaders on Android.
+
+> This setting is because Android supports GLES APIs but GLES does not support `NV_SHADER_EXTENSION`.
+
+**In line 50-58, you define and build the Taichi AOT module:**
+
+1. Create a Taichi AOT module, specifying its backend as OpenGL:
+
+```Python
+     m = ti.aot.Module(ti.opengl)
+```
+
+2. Add the required kernels `initialize` and `compute_force`, each with its own Ndarrays, to the module:
+
+```Python
+m.add_kernel(initialize, (pos, vel))
+
+m.add_kernel(compute_force, (pos, vel, force))
+```
+
+3. Specify a folder under your current working directory for holding the files that the module generates:
+
+```Python
+dir_name = 'nbody_aot'
+
+m.save(dir_name, '')
+```
+
+*The* *necessary* *compute* *shaders* *together with* *a* *JSON file* *appear* *under the specified directory.*
```

Comment:
nit: You don't need to star-ize every word...

Reply:
Oops. Something must go wrong when I did the copy&paste.